### DISCUSSION TITLE
"RuntimeError: 0. The ID token is not yet valid." can be refined

### DISCUSSION THREAD

**Discussion by jiasli** [AzureAD/microsoft-authentication-library-for-python#446] [Ideas]
Azure CLI users frequently see

```
RuntimeError: 0. The ID token is not yet valid. Current epoch = 1637158952.  The id_token was: {
```

All issues are summarized at 

- https://github.com/Azure/azure-cli/issues/20388

MSAL can refine the error message so that users can know what to do next:

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/9f78b6b9ed26fd4163ef24c49c648ce9a9de9990/msal/oauth2cli/oidc.py#L47-L50

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/9f78b6b9ed26fd4163ef24c49c648ce9a9de9990/msal/oauth2cli/oidc.py#L70-L71

**Comment by rayluo** [comment]
Thanks for sharing the real-world cases with us. How about we adjust those 2 messages like this? If you are fine with them, I can get them fixed in MSAL code base.

> err = "0. The ID token is not yet valid."

err = "0. Is the time on this computer correct? We seem to encounter an ID token issued in the future." 

> err = "9. The current time MUST be before the time represented by the exp Claim."

err = "9. Is the time on this computer correct? We seem to encounter an ID token already expired."

**Reply by jiasli** [reply]
Another occurrences of this error today: 

- https://github.com/Azure/azure-cli/issues/20825
- https://github.com/Azure/azure-cli/issues/20822
- https://github.com/Azure/azure-cli/issues/20816

This definitely works. I would personally prefer an imperative sentence:

> err = "0. The ID token is not yet valid. Make sure your computer's time is correctly synchronized."

A question like "Is the time on this computer correct?" is not obvious for what to do next.

BTW, Azure CLI actually has a whole error recommendation system:

1. Manually: https://github.com/Azure/azure-cli/blob/6751048cd85921081e3d033624f6109d7d6a01ec/src/azure-cli-core/azure/cli/core/util.py#L76-L83
2. AI-driven: https://github.com/Azure/azure-cli/blob/8f28e98186ad7db242b004da78c3f1ce7f94c11e/src/azure-cli-core/azure/cli/core/command_recommender.py#L122

+@chasewilson

**Reply by rayluo** [reply]
OK. I will adopt your suggestion on "error 0". How about the shortened "error 9"?

> err = "0. The ID token is not yet valid."

err = "0. The ID token is not yet valid. Make sure your computer's time is correctly synchronized."

> err = "9. The current time MUST be before the time represented by the exp Claim."

err = "9. The ID token already expires. Make sure your computer's time is correctly synchronized."

**Reply by jiasli** [reply]
I like the error message!

**Reply by rayluo** [reply]
Implemented and merged in PR #449
 (Note: It turns out a PR can not automatically close a Discussion topic. Let me manually close/lock it here.)
